src/libs/elektra/keyset.c

```diff
@@ -936,6 +936,125 @@ ssize_t ksAppend (KeySet * ks, const KeySet * toAppend)
 	return ks->size;
 }
 
+/**
+ * The core rename loop of ksRename()
+ */
+static size_t ksRenameInternal (KeySet * ks, size_t start, size_t end, const Key * root, const Key * newRoot)
+{
+	for (size_t it = start; it < end; ++it)
+	{
+		if (ks->array[it]->ksReference == 1)
+		{
+			// only referenced in this KeySet -> just override read-only flag
+			clear_bit (ks->array[it]->flags, KEY_FLAG_RO_NAME);
+		}
+		else
+		{
+			// key has other references -> dup in-place so we can safely rename it
+			Key * dup = keyDup (ks->array[it]);
+			keyDecRef (ks->array[it]);
+			dup->ksReference = 1;
+			keyDel (ks->array[it]);
+			ks->array[it] = dup;
+		}
+		keyReplacePrefix (ks->array[it], root, newRoot);
+		// lock key with new name
+		set_bit (ks->array[it]->flags, KEY_FLAG_RO_NAME);
+	}
+	return end - start;
+}
+
+/**
+ * Moves all keys below @p root to below @p newRoot
+ *
+ * Only keys below @p root will be modified. The rest of
+ * @p ks remains untouched.
+ *
+ * This functions is similar to the following snippet, but
+ * there are some differences.
+ *
+ * @code{.c}
+ * KeySet * toRename = ksCut (ks, root);
+ * for (elektraCursor cursor = 0; cursor < ksGetSize (toRename); cursor++)
+ * {
+ *     Key * cur = keyDup (ksAtCursor (ks, cursor));
+ *     keyReplacePrefix (cur, root, newRoot);
+ *     ksAppendKey (ks, cur);
+ * }
+ * ksDel (toRename);
+ * @endcode
+ *
+ * Firstly, the optimizations only work, if @p ks doesn't
+ * contain any keys below @p newRoot that aren't below @p root.
+ * If such keys exist, ksRename() will still work, but it will fall
+ * back to code similar to the for-loop above.
+ *
+ * The second difference is that ksRename() will modify the keys in
+ * @p ks directly, if they aren't referenced from anywhere else
+ * (if their reference count is 1 (see keyGetRef())).
+ * Normally, this shouldn't cause problems, but if you have a direct
+ * `Key *` pointer to a key in @p ks or hold a reference to some data
+ * within a key of @p ks, you may need to call keyIncRef() to ensure
+ * the key isn't modified.
+ *
+ * @param ks      the keyset to manipulate
+ * @param root    the old prefix that will be removed, must not be a cascading key
+ * @param newRoot the new prefix the will replace the old one, must not be a cascading key
+ *
+ * @retval -1 if any of @p ks, @p root, @p newRoot is NULL, or if @p root or @p newRoot are cascading keys
+ * @retval -2 if @p ks already contains keys below @p newRoot
+ * @retval  0 if @p ks contains no keys below @p root (and also not @p root itself)
+ * @returns   otherwise, the number of keys that have been renamed
+ */
+ssize_t ksRename (KeySet * ks, const Key * root, const Key * newRoot)
```

Comment:
We could add this function to `kdb.h` as well, but I put in in `kdbprivate.h` for now.

Mainly, because of the fact that it modifies keys with `key->ksReference == 1` in-place. This is nice, because you can avoid allocating new keys and sometimes even completely avoids re-sorting the KeySet, but it can easily lead to memory bugs. For example in a section like this:
https://github.com/ElektraInitiative/libelektra/blob/1e7ba555f1eeadd31f76d912230f0cf4a646910f/src/libs/elektra/kdb.c#L316-L322

However, apart from this the function is completely safe to use and could be useful in plugins that move keys around. So maybe putting it into `kdb.h` could make sense.